### PR TITLE
feat: Add more validation for user requests [DEV-3453]

### PR DESCRIPTION
**Pull request by anikitinDSR**
1. Cover controller's logic with try-catch logic to handle the errors without falling down
2. Refactor controllers and divide them into corresponding modules
3. Implement validation module as much independent is it can be. In case if we wanna to move it and reuse somewhere.
4. Add validation for:
  - DID
    - `cheqd` method
       - IndyStyle id
       - UUID
    - `key` method
  - DIDDocument:
    - VerificationMethod
    - Service
    - Authentication
    - Controller
  - Credential
  - Presentation
5. Add unit tests for all the main objects/structures
6. Fix small issues which was found while walking through the code

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-3453 Refactor validation of user requests](https://app.clickup.com/t/6600954/DEV-3453)

File: src/controllers/validator/also-known-as.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		if (!alsoKnownAs.description) {
+			return {
+				valid: false,
+				error: "Description field is required for AlsoKnownAs",
+			};
+		}
+		if (typeof alsoKnownAs.description !== 'string') {
+			return {
+				valid: false,
+				error: "Description field in AlsoKnownAs must be a string",
```
Comment (ankurdotb) [inline src/controllers/validator/also-known-as.ts]: Is this validator only meant to be used with resources? The rules here seem to be highly specific to our implementation of DID-linked resources. `alsoKnownAs` is also a property for DIDs, and this is allowed to be:

1. Just a string (typically a URI string, could be a DID or HTTP URI)
2. An array of strings (could either be DID or HTTP URIs)
3. An array of objects (our style for DID-Linked Resources)

File: src/controllers/validator/also-known-as.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	validate(alsoKnownAs: Validatable): IValidationResult {
+		alsoKnownAs = alsoKnownAs as AlsoKnownAs[];
+		if (!Array.isArray(alsoKnownAs)) {
+			return {
+				valid: false,
+				error: 'AlsoKnownAs must be an array.',
+			};
+		}
```
Comment (ankurdotb) [inline src/controllers/validator/also-known-as.ts]: Not strictly true, but it might be true in our ledger/SDK?

File: src/controllers/validator/also-known-as.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+		if (alsoKnownAs.length === 0) {
+			return {
+				valid: false,
+				error: 'AlsoKnownAs must have at least one entry.',
+			};
+		}
```
Comment (ankurdotb) [inline src/controllers/validator/also-known-as.ts]: Unsure about this. Are we saying this should _always_ be populated? This is not strictly true for DIDs, if it's also being used to validate for DIDs. For DID-Linked Resources, this is true **in our default implementation** but is optional. (Think of the Protobuf definitions.)

File: src/controllers/validator/controller.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		this.didValidator = didValidator;
+	}
+
+	validate(controller: Validatable): IValidationResult {
+		if (!Array.isArray(controller)) {
+			return {
+				valid: false,
+				error: 'Controller should be an array',
+			};
+		}
```
Comment (ankurdotb) [inline src/controllers/validator/controller.ts]: Not strictly true, can be a string

**Review by ankurdotb — COMMENTED** [review]
To discuss: is this repo the correct place to have these validations, or should they instead have been in:

1. ts-proto (definitely not)
2. SDK (maybe some low-level ones)
3. did-provider-cheqd (likeliest candidate for yes, and does it not have them already?)
4. This repo (_only_ conditions that can't/shouldn't be added to did-provider-cheqd)

File: src/controllers/validator/identifier.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+			}
+		} catch (e) {
+			return {
+				valid: false,
+				error: 'Indy-style DID identifier is not a valid Base58 string',
+			};
+		}
+		return { valid: true };
+	}
+}
```
[the remark this answers is not in the document]
Comment (ankurdotb) [reply src/controllers/validator/identifier.ts:33]: See https://did-registrar.cheqd.net/api-docs/#/Cheqd%20Helpers/get_did_document (which uses only SDK, not did-provider-cheqd). We don't generate the identifier on our ledger, so it must always be provided by a client. And in other "clients" like Veramo CLI + did-provider-cheqd OR in DID Registrar, we must be creating identifiers. So, shouldn't there be a function/method in SDK/did-provider-cheqd that helps create Indy-style identifiers?

Basically, credential-service is not the first client app to create DID identifiers, so what underlying logic can we rely on/replicate>

[the remark this answers is not in the document]
Comment (ankurdotb) [reply src/controllers/validator/identifier.ts:33]: PS - If there's no simple answer, then let's pick it up when we move the validation from here to SDK/did-provider-cheqd

[the remark this answers is not in the document]
Comment (anikitinDSR) [reply src/controllers/validator/identifier.ts:33]: Here I used the same validation as we do on ledger side. We just get the list of decoded bytes and comparing it with 16 (Indy-style length).

File: src/controllers/validator/service-create-request.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		const results = services.map((s) => this.validateEach(s));
+		if (results.some((r) => !r.valid)) {
+			return {
+				valid: false,
+				error:
+					'Service for Create DID Request has validation errors: ' + results.map((r) => r.error).join(', '),
+			};
+		}
+		return { valid: true };
+	}
```
[the remark this answers is not in the document]
Comment (anikitinDSR) [reply src/controllers/validator/service-create-request.ts:69]: We have 2 kind of Service structure to be honest on CredentialService side.
1. Came from `did-resolver` and it's used as a part of did document.
2. Another one is only our own and instead of `id` field it uses `idFragment` one and the main purpose of it - just provide the template for compiling the service section when the DID will be generated. For building DID we firstly make DID body generation and when the DID will be known we create service section
